File: www/universe/entity.js
({
	class: 'universe.entity',
	build: function() {
		$(this).mousedown(this.mousedown)

		return this
	},
	update: function(data) {
		var image = $('.entity-image', this)

		if (this.log) {
			console.log(data)
			this.log = false
		}

		this.id = data.id

		if (image.attr('src') != data.imaging.image) {
			image.attr('src', data.imaging.image)
		}

		if (this.x != data.shape.anchor.x || this.y != data.shape.anchor.y) {
			this.x = data.shape.anchor.x
			this.y = data.shape.anchor.y

			$(this).css({
				left: data.shape.anchor.x,
				top: data.shape.anchor.y
			})
		}

		if (this.dx != data.shape.dimension.dx || this.dy != data.shape.dimension.dy) {
			this.dx = data.shape.dimension.dx
			this.dy = data.shape.dimension.dy

			image.css({
				width: data.shape.dimension.dx,
				height: data.shape.dimension.dy
			})
		}

		if (data.player) {
			var entity = this
			ospokemon.element.get('universe/nameplate').then(function(el) {
				if (!entity.player) {
					entity.player = el.build(entity)
				}

				entity.player.update(data.player)
			})
		}
	},
	mousedown: function(e) {
		if (e.button == 2) {
			ospokemon.websocket.Send('Click.Entity', {
				entity: parseInt(this.id)
			})
		}
		return false
	}
})

File: json/entity.go
package json

import (
	"github.com/ospokemon/ospokemon/game"
)

func Entity(e *game.Entity) (string, map[string]interface{}) {
	data := map[string]interface{}{
		"id":    e.Id,
		"shape": e.Shape.Snapshot(),
	}

	for _, part := range e.Parts {
		if player, ok := part.(*game.Player); ok {
			key, partData := Player(player)
			data[key] = partData
		} else if imaging, ok := part.(*game.Imaging); ok {
			key, partData := Imaging(imaging)
			data[key] = partData
		}
	}

	return "entity", data
}
